Shared/configurations/ConfigurationController.cpp

```diff
@@ -0,0 +1,373 @@
+  });
+}
+
+void ConfigurationController::cancel(int index)
+{
+  Q_UNUSED(index); // TODO: supporting multiple downloads at a time will require additional management
+  if (m_networkReply)
+  {
+    if (m_networkReply != nullptr)
+    {
+      if (m_networkReply->isOpen())
+      {
+        m_configurationListModel->cancel(index);
+        m_aborted = true;
+        m_downloading = false;
+        m_networkReply->abort();
+        m_networkReply->deleteLater();
+        m_networkReply = nullptr;
+      }
+    }
+  }
+}
+
+void ConfigurationController::remove(int index)
+{
+  // make extracted packages removeable from device
+  Q_UNUSED(index);
+}
+
+void ConfigurationController::storeConfigurations()
+{
+  QJsonArray array;
+  for (const auto& cfg : *m_configurationListModel)
+  {
+    QJsonObject obj;
+    obj["selected"] = cfg.selected();
+    obj["name"] = cfg.name();
+    obj["url"] = cfg.url();
+    array.append(obj);
+  }
+  QJsonDocument doc{array};
+  QFile file{DsaUtility::configurationsFilePath()};
+  if (file.open(QIODevice::WriteOnly))
+    file.write(doc.toJson());
+}
+
+bool ConfigurationController::createDefaultConfigurationsFile()
+{
+  // create a single entry for the default configuration
+  QJsonObject obj;
+  obj["selected"] = true;
+  obj["name"] = "Default";
+  obj["url"] = URL_DEFAULT_DOWNLOAD;
+
+  // create a new document as a top level array and append the default object
+  QJsonArray array;
+  array.append(obj);
+  QJsonDocument doc{array};
+
+  // write out the document to disk
+  QFile file{DsaUtility::configurationsFilePath()};
+  if (!file.open(QIODevice::WriteOnly))
+    return false;
+  file.write(doc.toJson());
+  return true;
+}
+
+QString ConfigurationController::toolName() const
+{
+  return QStringLiteral("configurations");
+}
+
+void ConfigurationController::setProperties(const QVariantMap& properties)
+{
+  Q_UNUSED(properties);
+}
+
+bool ConfigurationController::downloading()
+{
+  return m_downloading;
+}
+
+QAbstractListModel* ConfigurationController::configurations() const
+{
+  return m_configurationListModel;
+}
+
+void ConfigurationController::downloadProgress(quint64 bytesReceived, quint64 bytesTotal)
+{
+  setPercentComplete(bytesReceived * 1.0 / bytesTotal * 1.0 * 100);
+}
+
+void ConfigurationController::readyRead()
+{
+  // make sure the download file can be opened for appending
+  QFile file(m_downloadFileName);
+  if (!file.open(QIODevice::Append))
+  {
+    m_aborted = true;
+    return;
+  }
+
+  // append the latest chunk to the download file in progress
+  file.write(m_networkReply->readAll());
+}
+
+void ConfigurationController::finished()
+{
+  // turn off the downloading state and clean up any aborted download files remaining
+  m_downloading = false;
+  if (m_aborted)
+  {
+    QFile fileAbort(m_downloadFileName);
+    if (fileAbort.exists())
+      fileAbort.remove();
+    m_aborted = false;
```

Comment:
There are a few of these. I won't comment on them all to keep the review comments to a minimum, but please add a line after if statement conditions.
```suggestion
    if (fileAbort.exists())
      fileAbort.remove();

    m_aborted = false;
```

IMO it can hide conditional logic when someone new is reading through the code. Alternately you can always use braces for single-statement condition blocks.

Reply:
Agreed, this should be avoided. I've added extra line for clarity and updated comments where appropriate.